File: api/src/dnc/check.ts
// D05 — DNC hot-path check (TS / API side).
//
// Algorithm per PLAN §2.1:
//  1. Validate phone format.
//  2. Pipeline BF.EXISTS across sources (one RTT).
//  3. On all-negative → return IsDNC=false (typical case).
//  4. For positive sources → MySQL confirmation query.
//  5. Return result + emit metrics.

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyRedis = any;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyPrisma = any;
import { parsePhoneNumberFromString } from "libphonenumber-js";
import {
  type CheckRequest,
  type CheckResult,
  type DncSource,
  sortSourcesByPriority,
} from "./types.js";
import { bloomMexistsPipeline } from "./bloom.js";

// ── MySQL confirmation ────────────────────────────────────────────────────────

interface DncRow {
  source: string;
  state: string;
  campaign_id: string;
}

async function confirmMySQL(
  prisma: AnyPrisma,
  phone: string,
  tenantId: number,
  campaignId: string | undefined,
  leadState: string | undefined,
  positiveSources: DncSource[],
): Promise<DncSource[]> {
  // Build raw SQL per PLAN §2.1 step 4
  const sourceList = positiveSources.map((s) => `'${s}'`).join(",");
  const campId = campaignId ?? "__GLOBAL__";
  const st = leadState ?? "__";

   
  const rows: DncRow[] = await prisma.$queryRawUnsafe(
    `SELECT source, state, campaign_id FROM dnc
     WHERE phone_e164 = ?
       AND tenant_id IN (?, 0)
       AND source IN (${sourceList})
       AND (
             source = 'federal'
          OR source = 'litigator'
          OR (source = 'internal' AND tenant_id = ?
                AND campaign_id IN ('__GLOBAL__', ?))
          OR (source = 'state' AND tenant_id = 0
                AND state IN (?, '__'))
       )
       AND (expires_at IS NULL OR expires_at > NOW())
     LIMIT 4`,
    phone,
    tenantId,
    tenantId,
    campId,
    st,
  );

   
  return [...new Set(rows.map((r) => r.source as DncSource))];
}

// ── Normalize phone ───────────────────────────────────────────────────────────

function normalizePhone(raw: string): string | null {
  // Accept E.164 or local; re-validate
  const parsed = parsePhoneNumberFromString(raw, "US");
  if (!parsed || !parsed.isValid()) return null;
  return parsed.format("E.164"); // "+1XXXXXXXXXX"
}

// ── Main Check ────────────────────────────────────────────────────────────────

export async function dncCheck(
  redis: AnyRedis,
  prisma: AnyPrisma,
  req: CheckRequest,
): Promise<CheckResult> {
  const start = process.hrtime.bigint();

  // Step 1: validate phone
  const phone = normalizePhone(req.phoneE164);
  if (!phone) {
    return {
      isDnc: true,
      sources: [],
      latencyMicros: Number(process.hrtime.bigint() - start) / 1000,
      bloomFalsePositive: false,
      reason: "malformed",
    };
  }

  // Step 2: Bloom pipeline
  const bloomHits = await bloomMexistsPipeline(
    redis,
    req.sources,
    req.tenantId,
    phone,
  );

  const positiveSources = req.sources.filter((s) => bloomHits.get(s) === true);

  const latencyAfterBloom = Number(process.hrtime.bigint() - start) / 1000;

  // Step 3: all-negative → fast path
  if (positiveSources.length === 0) {
    return {
      isDnc: false,
      sources: [],
      latencyMicros: latencyAfterBloom,
      bloomFalsePositive: false,
    };
  }

  // Step 4: MySQL confirmation
  let confirmedSources: DncSource[] = [];
  try {
    confirmedSources = await confirmMySQL(
      prisma,
      phone,
      req.tenantId,
      req.campaignId,
      req.leadState,
      positiveSources,
    );
  } catch {
    // MySQL unreachable on positive bloom — fail-closed (PLAN §1.5)
    confirmedSources = positiveSources;
  }

  const latencyTotal = Number(process.hrtime.bigint() - start) / 1000;
  const bloomFp = positiveSources.length > 0 && confirmedSources.length === 0;

  // Step 5: result
  if (confirmedSources.length === 0) {
    return {
      isDnc: false,
      sources: [],
      latencyMicros: latencyTotal,
      bloomFalsePositive: bloomFp,
    };
  }

  return {
    isDnc: true,
    sources: sortSourcesByPriority(confirmedSources),
    latencyMicros: latencyTotal,
    bloomFalsePositive: false,
  };
}

File: dialer/internal/dnc/check.go
package dnc

import (
	"context"
	"database/sql"
	"regexp"
	"sort"
	"time"

	"github.com/redis/go-redis/v9"
)

// e164RE validates E.164 phone format.
var e164RE = regexp.MustCompile(`^\+[1-9]\d{1,14}$`)

// Checker is the DNC check primitive used by E01 hopper filler and T04 originate gate.
// It holds references to Valkey and MySQL; no network round-trips beyond those two.
type Checker struct {
	rdb     redis.UniversalClient
	db      *sql.DB
	metrics *Metrics
}

// NewChecker creates a Checker. metrics may be nil (no-op).
func NewChecker(rdb redis.UniversalClient, db *sql.DB, metrics *Metrics) *Checker {
	return &Checker{rdb: rdb, db: db, metrics: metrics}
}

// Check implements the DNC hot-path algorithm (PLAN §2.1).
func (c *Checker) Check(ctx context.Context, req CheckRequest) CheckResult {
	start := time.Now()

	// Step 1: validate phone
	if !e164RE.MatchString(req.PhoneE164) {
		return CheckResult{
			IsDNC:         true,
			LatencyMicros: time.Since(start).Microseconds(),
			Reason:        "malformed",
		}
	}

	// Step 2: Bloom pipeline
	bloomHits := bloomMexists(ctx, c.rdb, req.Sources, req.TenantID, req.PhoneE164)

	// Collect positive sources
	var positiveSources []Source
	for _, src := range req.Sources {
		if bloomHits[src] {
			positiveSources = append(positiveSources, src)
		}
	}

	// Step 3: all-negative → fast path
	if len(positiveSources) == 0 {
		c.recordCheck(req.Sources, "miss", time.Since(start))
		return CheckResult{
			IsDNC:         false,
			LatencyMicros: time.Since(start).Microseconds(),
		}
	}

	// Step 4: MySQL confirmation
	confirmed, err := confirmMySQL(
		ctx, c.db,
		req.PhoneE164, req.TenantID, req.CampaignID, req.LeadState,
		positiveSources,
	)
	if err != nil {
		// MySQL unavailable → fail-closed (PLAN §1.5)
		c.recordCheck(positiveSources, "fail_closed", time.Since(start))
		return CheckResult{
			IsDNC:         true,
			Sources:       positiveSources,
			LatencyMicros: time.Since(start).Microseconds(),
		}
	}

	latency := time.Since(start)

	if len(confirmed) == 0 {
		// Bloom false positive
		c.recordFalsePositives(positiveSources)
		c.recordCheck(req.Sources, "false_positive", latency)
		return CheckResult{
			IsDNC:              false,
			BloomFalsePositive: true,
			LatencyMicros:      latency.Microseconds(),
		}
	}

	c.recordCheck(confirmed, "hit", latency)
	return CheckResult{
		IsDNC:         true,
		Sources:       sortSources(confirmed),
		LatencyMicros: latency.Microseconds(),
	}
}

// sortSources sorts by priority (PLAN §2.3): internal > litigator > state > federal.
func sortSources(sources []Source) []Source {
	out := make([]Source, len(sources))
	copy(out, sources)
	sort.Slice(out, func(i, j int) bool {
		return sourcePriority[out[i]] > sourcePriority[out[j]]
	})
	return out
}

func (c *Checker) recordCheck(sources []Source, outcome string, latency time.Duration) {
	if c.metrics == nil {
		return
	}
	for _, src := range sources {
		c.metrics.CheckTotal.WithLabelValues(string(src), outcome).Inc()
		c.metrics.CheckLatency.WithLabelValues(string(src)).Observe(latency.Seconds())
	}
}

func (c *Checker) recordFalsePositives(sources []Source) {
	if c.metrics == nil {
		return
	}
	for _, src := range sources {
		c.metrics.FalsePositiveTotal.WithLabelValues(string(src)).Inc()
	}
}
